tests/importer/importer.cpp

```diff
@@ -0,0 +1,762 @@
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(size_t(1), importer->warningCount());
+    EXPECT_EQ(warningMessage, importer->warning(0)->description());
+}
+
+TEST(Model, missingUnitsFromImportOfCnTerms)
+{
+    // This test is intended to show that parsing a model and importing
+    // it have the same end result.
+    auto validator = libcellml::Validator::create();
+    auto model = libcellml::Model::create("model_from_imports");
+    auto c = libcellml::Component::create("c");
+    auto importer = libcellml::Importer::create();
+
+    auto imp = libcellml::ImportSource::create();
+    imp->setUrl("units_in_cn.cellml");
+
+    c->setImportReference("myComponent");
+    c->setImportSource(imp);
+    model->addComponent(c);
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+
+    importer->resolveImports(model, resourcePath());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+    model = importer->flatten(model);
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->errorCount());
+}
+
+TEST(Model, importingComponentWithCnUnitsThatAreAlreadyDefinedInImportingModel)
+{
+    const std::string e =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"myModel\">\n"
+        "  <units name=\"myUnitsThatIUse\">\n"
+        "    <unit units=\"metre\"/>\n"
+        "  </units>\n"
+        "  <units name=\"myUnitsThatIUse_1\">\n"
+        "    <unit units=\"second\"/>\n"
+        "  </units>\n"
+        "  <component name=\"c\">\n"
+        "    <variable name=\"a\" units=\"second\"/>\n"
+        "    <math xmlns=\"http://www.w3.org/1998/Math/MathML\" xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\">\n"
+        "      <apply>\n"
+        "        <eq/>\n"
+        "        <ci>a</ci>\n"
+        "        <cn cellml:units=\"myUnitsThatIUse_1\">1</cn>\n"
+        "      </apply>\n"
+        "    </math>\n"
+        "  </component>\n"
+        "</model>\n";
+
+    const std::string in =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"myModel\">\n"
+        "  <units name=\"myUnitsThatIUse\">\n"
+        "    <unit units=\"second\"/>\n"
+        "  </units>\n"
+        "   <component name=\"myComponent\">\n"
+        "       <variable name=\"a\" units=\"second\"/>\n"
+        "       <math xmlns=\"http://www.w3.org/1998/Math/MathML\" xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\">\n"
+        "           <apply>"
+        "             <eq/>\n"
+        "             <ci>a</ci>\n"
+        "             <cn cellml:units=\"myUnitsThatIUse\">1</cn>\n"
+        "           </apply>\n"
+        "       </math>\n"
+        "   </component>\n"
+        "</model>";
+
+    // Create the model by parsing the string above.
+    auto parser = libcellml::Parser::create();
+    auto importedModel = parser->parseModel(in);
+    auto importer = libcellml::Importer::create();
```

Comment:
This really ought to be declared when actually needed, i.e. just before line 236.

Reply:
I prefer to do it at the beginning - I think it's mixed up throughout the code as to where things are declared.

Reply:
FYI, that's old C style of declaring variables... :)